router/h2/src/main/scala/io/buoyant/router/h2/ClassifierFilter.scala

```diff
@@ -1,50 +1,155 @@
+
+  @inline private[this] def classifyStream(
+    req: Request,
+    rep: Response,
+    frame: Option[Frame] = None
+  ): String = {
+    val rr = H2ReqRepFrame(req, Return((rep, frame.map(Return(_)))))
+    val respClass = classifier.streamClassifier(rr)
+    successHeader(respClass)
+  }
 
   def apply(req: Request, svc: Service[Request, Response]): Future[Response] = {
-    svc(req).map { rep =>
-      successClass(ReqRep(req, Return(rep))) match {
-        case Some(success) =>
-          rep.headers.set(ClassifierFilter.SuccessClassHeader, success.toString)
-        case None =>
-      }
-      rep
+    svc(req).map { rep: Response =>
+      classifyEarly(H2ReqRep(req, rep))
+        .map { success =>
+          // classify early - response class goes in headers
+          rep.headers.set(SuccessClassHeader, success)
+          rep
+        }
+        .getOrElse {
+          // if the early classification attempt is not defined, attempt
+          // late classification on the last frame in the response stream...
+          if (rep.stream.isEmpty) {
+            val success = classifyStream(req, rep)
+            rep.headers.set(SuccessClassHeader, success)
+            rep
+          } else {
+            // flatMap over the frames in the response stream, since we may
+            // need to add a new Trailers frame to the end of the stream.
+            val stream = rep.stream.flatMap {
+              case frame: Trailers =>
+                // if the current frame is a Trailers frame, we've reached
+                // the end of the stream. just add the  success class header
+                // to it.
+                val success = classifyStream(req, rep, Some(frame))
+                frame.set(SuccessClassHeader, success)
+                frame.set("te", "trailers")
+                Seq(frame)
+              case frame: Frame.Data if frame.isEnd =>
+                // if the current frame is a data frame with the end of stream
+                // flag, then we need to send this frame followed by a new
+                // Trailers frame
+                val success = classifyStream(req, rep, Some(frame))
+                // since this frame has the end of stream flag, we need to
+                // replace it with a new data frame without that flag.
+                // otherwise, the trailers we add to the stream will never
+                // be reached.
+                val frame2 = Frame.Data(frame.buf, eos = false)
+                // release the old data frame, since it will be replaced.
+                frame.release()
+                val trailers = Trailers(
+                  SuccessClassHeader -> success,
+                  "te" -> "trailers"
+                )
+                Seq(frame2, trailers)
+              case frame =>
+                // if the current frame is not an end frame, just keep going...
+                Seq(frame)
+            }
+            Response(rep.headers, stream)
```

Comment:
I think we need to add `te: trailers` to the headers here.  At this point we're committed to adding a trailers frame (if one doesn't already exist).